lib/src/async_cache.dart

```diff
@@ -43,70 +48,76 @@ class AsyncCache<T> {
   /// executed at most once concurrently. This is useful for requests which data
-  /// is updated frequently but the cost of making multiple requests without
-  /// previous completing yet is prohibitive or unwanted.
+  /// is updated frequently but stale data is acceptable.
   ///
   /// For _testing_, a custom `now` function may be provided.
   factory AsyncCache.ephemeral({DateTime now()}) =>
       new AsyncCache._(Duration.ZERO);
 
   // Prevent inheritance. This allows using a redirecting factory constructor
   // with different implementations in the future without a breaking API change.
-  AsyncCache._(this._duration, {DateTime now(): _defaultNow,})
+  AsyncCache._(
+    this._duration, {
+    DateTime now(): _defaultNow,
+  })
       : _getNow = now;
 
   /// Returns a cached value or runs [callback] to compute a new one.
   ///
   /// If [callback] has been run recently enough, returns its previous return
   /// value. Otherwise, runs [callback] and returns its new return value.
   Future<T> fetch(Future<T> callback()) async {
+    if (_cachedStreamSplitter != null) {
+      throw new StateError('Previously used to cache via `fetchStream`');
+    }
     _invalidateWhenStale();
-    var result = await (_cachedValueFuture ??= callback());
+    if (_cachedValueFuture == null) {
+      _cachedValueFuture = callback();
+    }
+    var result = await _cachedValueFuture;
     _cachedExpiration ??= _getNow().add(_duration);
     return result;
   }
 
   /// Returns a cached stream or runs [callback] to compute a new stream.
   ///
-  /// If [callback] has been run recently enough, returns its previous return
-  /// value. Otherwise, runs [callback] and returns its new return value.
+  /// If [callback] has been run recently enough, returns a copy of its previous
+  /// return value. Otherwise, runs [callback] and returns its new return value.
   ///
   /// If a stream is currently being fetched, waits until the _done_ event and
   /// then returns the cached value.
-  Stream<T> fetchStream(Stream<T> callback()) async* {
-    _invalidateWhenStale();
-    if (_fetchingStream != null) {
-      await _fetchingStream;
+  Stream<T> fetchStream(Stream<T> callback()) {
+    if (_cachedValueFuture != null) {
+      throw new StateError('Previously used to cache via `fetch`');
     }
-    if (_cachedStreamValues != null) {
-      yield* new Stream<T>.fromIterable(_cachedStreamValues);
-      return;
+    if (_isInvalidated) {
+      _cachedStreamSplitter.close();
+      invalidate();
     }
-    var fetching = new Completer();
-    _fetchingStream = fetching.future;
-    var values = <T>[];
-    await for (var result in callback()) {
-      values.add(result);
-      yield result;
+    if (_cachedStreamSplitter == null) {
+      _cachedStreamSplitter = new StreamSplitter<T>(callback());
+      _cachedStreamSplitter.split().last.then((_) {
+        _cachedExpiration = _getNow().add(_duration);
+      });
```

Comment:
`.last` will throw if the stream emits an error, and then that error will be unhandled. You could use `.listen(null, onDone: ...)` to avoid that.

That said, eagerly listening here breaks some of the laziness benefits of streams and `StringSplitter`'s nice cancel/pause handling. If you transform the input stream, you can avoid that:

```dart
new StreamSplitter(callback().transform(
    new StreamTransformer.fromHandlers(handleDone: (sink) {
  _cachedExpiration = _getNow().add(_duration);
})))
```